### ISSUE TITLE
Brian hears won't work with python 3?

### ISSUE THREAD

**Issue by choldgraf** [brian-team/brian2#855]
It seems like using `brian.hears` in Brian 2 requires having an installation of Brian 1...but if Brian 1 is not compatible with Python 3, doesn't that mean that you can't use `brian.hears` in Python 3, or am I misunderstanding?

**Comment by thesamovar** [comment]
Unfortunately this is correct.

**Comment by choldgraf** [comment]
Hmm - and you think it's a non-trivial amount of work to port brian hears functionality to python 3? I could try a PR at some point if it's something on the order of many hours...I like Brian hears!

**Comment by thesamovar** [comment]
I haven't thought about what it would involve, it might not be too bad. The actual conversion to Python 3 syntax etc. would probably be quite straightforward and could mostly be done using the 2to3 tool. The trickier thing is that you'd want to separate Brian Hears from the rest of Brian 1, because you wouldn't want to convert the whole code base of Brian 1 to work in Python 3. I haven't looked at the code base of Brian Hears lately, so I don't remember how tightly coupled it is with the rest of Brian 1. Funnily enough, I was actually thinking about this recently because I'm considering moving to Python 3 as my standard version (yeah I'm a slow adopter), and BH is the main thing holding me back.

If you want to have a go we'd definitely be up for providing support.

**Comment by choldgraf** [comment]
Sounds good, I can take a look at the codebase. How intertwined is hears
with brian? I switched to python 3 a few months back, when I realized that
Berkeley had basically stopped teaching students using python 2  :-P

**Comment by thesamovar** [comment]
I don't remember. It's certainly intertwined with the Brian 1 units system, but possibly not so much for the rest. Might be easier just to try to take the code and rewrite to work with the Brian 2 units system rather than do a Py3 version of the Brian 1 units system, not sure. If you do decide to give it a try, it might be easiest to release it as a separate package rather than do a PR and put it in Brian 2, because the review process for Brian 2 is a bit more rigorous in terms of testing. An option is to transfer that package into the Brian GitHub organisation when it's done.

**Comment by choldgraf** [comment]
Not a bad idea... Lemme think about it and I'll ping you if I start working
on something!

**Comment by thesamovar** [comment]
Great! I'll be away (and actually off the internet) July 14-25, but Marcel might be around to help if I'm not there. He's also away at the moment but at a work thing I think, so might be able to find some time to help too.

**Comment by mstimberg** [comment]
Unfortunately I will not be around the next weeks either... Just for context: we had discussions a long time ago about completely re-writing Brian Hears to use code generation (i.e. to make it a proper Brian2Hears) -- see the https://github.com/brian-team/brian2hears repository for early attempts). This would be a really major undertaking, though, so that's why we introduced the "bridge" to use Brian hears from Brian2 (also see #139) as a stopgap solution.
Not sure whether we should have a "quick&dirty" version of Brian2Hears by simply "fixing" Brian Hears to work with Brian 2, or still wait for a laborious re-write with Brian2's code generation system.

**Comment by thesamovar** [comment]
I think it might take years before we get round to making a proper Brian 2 version of Brian Hears so a quick and dirty version is an excellent idea (and I'll help making it work).

**Comment by sammosummo** [comment]
@choldgraf and @thesamovar, did you guys make any progress on the quick-'n'-dirty version?

**Comment by thesamovar** [comment]
Not yet. I still think it would be worth doing it's just a matter of someone finding the time.

**Comment by sammosummo** [comment]
OK. I suspect I probably don’t have the know-how to do it myself :(

For my own purposes I’m mainly interested in the simple sound manipulation and synthesis stuff like tones etc. Doing this from scratch in Python is easy but I really enjoyed the ease of using real units (specifying slices in s or ms; changing the amplitude in dB; etc.). Do you have any suggestions on how to port just these functions to Python 3? For instance, could you guess at how difficult would it be to get just those working with the unit structure from brian2?

**Comment by mstimberg** [comment]
Hi @sammosummo, did you (or anyone else) try to put Brian (or only Brian Hears) through the 2to3.py module? Maybe it "just works" and then you could use it with the "Brian Hears bridge":
https://brian2.readthedocs.io/en/stable/introduction/brian1_to_2/brian1hears_bridge.html

**Comment by thesamovar** [comment]
I think you could use the units system from Brian 2 (which works in py3) and just copy across the Sound class and it shouldn't be too difficult to adapt. Unfortunately, I think that the slicing in seconds/ms rather than samples won't work with either the units system of Brian 2 or because of changes in py3 (I don't remember which was the source of the problem). You could modify it so that you have an equivalent method though.

**Comment by choldgraf** [comment]
nope, I haven't made any progress on it. FWIW this will become more important soon, since many major python packages are now dropping support for 2.7 in new features (e.g. the latest Matplotlib 3.0 release no longer supports Python 2.x)

**Comment by thesamovar** [comment]
Yep, that's on our minds.

**Comment by choldgraf** [comment]
glad to see this one closed :-)

**Comment by sammosummo** [comment]
Awesome, looking forward to trying it out!
